File: frontend/src/core/fieldValues/AccountType.ts
import FieldValue from "@core/fieldValues/FieldValue";

/**
 * AccountType field value that represents the different types of accounts.
 */
class AccountType extends FieldValue {
  /**
   * A Standard account represents a standard checking or savings account.
   */
  public static readonly Standard: AccountType = new AccountType("Standard");

  /**
   * A Debt account represents a credit card or loan account.
   */
  public static readonly Debt: AccountType = new AccountType("Debt");

  /**
   * An Investment account represents a retirement or brokerage account.
   */
  public static readonly Investment: AccountType = new AccountType(
    "Investment",
  );

  /**
   * Collection of all AccountTypes.
   */
  public static override readonly Collection = [
    AccountType.Standard,
    AccountType.Debt,
    AccountType.Investment,
  ];
}

export default AccountType;

File: backend/Domain/Accounts/AccountType.cs
namespace Domain.Accounts;

/// <summary>
/// Enum representing the different Account types
/// </summary>
public enum AccountType
{
    /// <summary>
    /// Standard Account
    /// </summary>
    Standard,

    /// <summary>
    /// Credit Card Account
    /// </summary>
    CreditCard,

    /// <summary>
    /// Investment Account
    /// </summary>
    Investment,

    /// <summary>
    /// Debt Account
    /// </summary>
    Debt,

    /// <summary>
    /// Retirement Account
    /// </summary>
    Retirement,

    /// <summary>
    /// Escrow Account
    /// </summary>
    Escrow,
}

/// <summary>
/// Extension methods for AccountType
/// </summary>
public static class AccountTypeExtensions
{
    /// <summary>
    /// Determines if an AccountType is tracked
    /// </summary>
    public static bool IsTracked(this AccountType accountType) => accountType switch
    {
        AccountType.Standard => true,
        AccountType.CreditCard => true,
        AccountType.Investment => true,
        AccountType.Debt => false,
        AccountType.Retirement => false,
        AccountType.Escrow => false,
        _ => throw new ArgumentOutOfRangeException(nameof(accountType), $"Not expected account type value: {accountType}"),
    };

    /// <summary>
    /// Determines if an AccountType is a debt account
    /// </summary>
    public static bool IsDebt(this AccountType accountType) => accountType switch
    {
        AccountType.Standard => false,
        AccountType.CreditCard => true,
        AccountType.Investment => false,
        AccountType.Debt => true,
        AccountType.Retirement => false,
        AccountType.Escrow => false,
        _ => throw new ArgumentOutOfRangeException(nameof(accountType), $"Not expected account type value: {accountType}"),
    };
}